File: src/shared/tree.ts
import { z } from 'zod';

export const nodeKindSchema = z.enum([
  'connection',
  'database',
  'schema',
  'table',
  'view',
  'matview',
  'function',
  'sequence',
  'routine',
  'column',
]);
export type NodeKind = z.infer<typeof nodeKindSchema>;

export interface PathSegment {
  kind: NodeKind;
  name: string;
}

export interface NodePath {
  connectionId: string;
  segments: PathSegment[];
}

// D6: node paths are encoded as `kind:name` segments joined by `/`, with `name` percent-encoded.
// The encoded form excludes the connection id (metadata_cache's key is already (connection_id,
// path)). Example: 'schema:public/table:order%2Fitems'. `encodePath([])` is '' — the root.
export function encodePath(segments: PathSegment[]): string {
  return segments.map((s) => `${s.kind}:${encodeURIComponent(s.name)}`).join('/');
}

// `decodePath` throws on an unknown kind or a malformed segment; callers treat that as a corrupt
// cache row and drop it.
export function decodePath(connectionId: string, encoded: string): NodePath {
  if (encoded === '') return { connectionId, segments: [] };
  const segments = encoded.split('/').map((part) => {
    const colon = part.indexOf(':');
    if (colon <= 0) throw new Error(`malformed path segment "${part}"`);
    const kind = part.slice(0, colon);
    const parsed = nodeKindSchema.safeParse(kind);
    if (!parsed.success) throw new Error(`unknown node kind "${kind}"`);
    return { kind: parsed.data, name: decodeURIComponent(part.slice(colon + 1)) };
  });
  return { connectionId, segments };
}

export function pathParent(encoded: string): string | null {
  if (encoded === '') return null;
  const last = encoded.lastIndexOf('/');
  return last === -1 ? '' : encoded.slice(0, last);
}

export function pathTail(encoded: string): PathSegment | null {
  if (encoded === '') return null;
  const last = encoded.lastIndexOf('/');
  const part = last === -1 ? encoded : encoded.slice(last + 1);
  const colon = part.indexOf(':');
  if (colon <= 0) return null;
  const parsed = nodeKindSchema.safeParse(part.slice(0, colon));
  if (!parsed.success) return null;
  return { kind: parsed.data, name: decodeURIComponent(part.slice(colon + 1)) };
}

export const treeNodeSchema = z.object({
  kind: nodeKindSchema,
  name: z.string(),
  path: z.string(),
  hasChildren: z.boolean(),
  detail: z.string().optional(),
  badges: z.array(z.string()).optional(),
});
export type TreeNode = z.infer<typeof treeNodeSchema>;

export const columnMetaSchema = z.object({
  name: z.string(),
  position: z.number(),
  dataType: z.string(),
  nullable: z.boolean(),
  defaultExpr: z.string().nullable(),
  isPrimaryKey: z.boolean(),
  comment: z.string().nullable(),
});
export type ColumnMeta = z.infer<typeof columnMetaSchema>;

export const indexMetaSchema = z.object({
  name: z.string(),
  columns: z.array(z.string()),
  unique: z.boolean(),
  primary: z.boolean(),
  method: z.string().nullable(),
});
export type IndexMeta = z.infer<typeof indexMetaSchema>;

export const foreignKeyMetaSchema = z.object({
  name: z.string(),
  columns: z.array(z.string()),
  referencedPath: z.string(),
  referencedColumns: z.array(z.string()),
  onDelete: z.string().nullable(),
  onUpdate: z.string().nullable(),
});
export type ForeignKeyMeta = z.infer<typeof foreignKeyMetaSchema>;

export const objectMetaSchema = z.object({
  path: z.string(),
  kind: nodeKindSchema,
  name: z.string(),
  qualifiedName: z.string(),
  columns: z.array(columnMetaSchema),
  primaryKey: z.array(z.string()).nullable(),
  foreignKeys: z.array(foreignKeyMetaSchema),
  referencedBy: z.array(foreignKeyMetaSchema),
  indexes: z.array(indexMetaSchema),
  rowEstimate: z.number().nullable(),
  comment: z.string().nullable(),
});
export type ObjectMeta = z.infer<typeof objectMetaSchema>;

File: apps/kira-studio/internal/storage/model/tree.go
package model

import (
	"fmt"
	"strings"
)

// nodeKinds mirrors domain/tree.ts's nodeKindSchema — every project-tree level across all eleven
// adapters, confirmed against that file for P55 §4.6.
var nodeKinds = map[string]bool{
	"connection": true, "database": true, "schema": true, "table": true, "view": true,
	"matview": true, "function": true, "sequence": true, "column": true, "collection": true,
	"namespace": true, "key": true, "topic": true, "partition": true, "consumerGroup": true,
	"queue": true, "bucket": true, "prefix": true, "object": true, "exchange": true,
}

// ValidNodeKind mirrors domain/tree.ts's nodeKindSchema.
func ValidNodeKind(v string) bool { return nodeKinds[v] }

// PathSegment mirrors domain/tree.ts's PathSegment.
type PathSegment struct {
	Kind string `json:"kind"`
	Name string `json:"name"`
}

// NodePath mirrors domain/tree.ts's NodePath.
type NodePath struct {
	ConnectionID string        `json:"connectionId"`
	Segments     []PathSegment `json:"segments"`
}

// TreeNode mirrors domain/tree.ts's treeNodeSchema.
type TreeNode struct {
	Kind        string   `json:"kind"`
	Name        string   `json:"name"` // the raw identifier, used to build SQL and to copy
	Path        string   `json:"path"` // encoded, relative to the connection
	HasChildren bool     `json:"hasChildren"`
	Detail      *string  `json:"detail,omitempty"`
	Badges      []string `json:"badges,omitempty"`
}

// ColumnMeta mirrors domain/tree.ts's columnMetaSchema.
type ColumnMeta struct {
	Name         string  `json:"name"`
	Position     int     `json:"position"`
	DataType     string  `json:"dataType"`
	Nullable     bool    `json:"nullable"`
	DefaultExpr  *string `json:"defaultExpr"`
	IsPrimaryKey bool    `json:"isPrimaryKey"`
	Comment      *string `json:"comment"`
}

// IndexMeta mirrors domain/tree.ts's indexMetaSchema.
type IndexMeta struct {
	Name    string   `json:"name"`
	Columns []string `json:"columns"`
	Unique  bool     `json:"unique"`
	Primary bool     `json:"primary"`
	Method  *string  `json:"method"`
}

// ForeignKeyMeta mirrors domain/tree.ts's foreignKeyMetaSchema.
type ForeignKeyMeta struct {
	Name              string   `json:"name"`
	Columns           []string `json:"columns"`
	ReferencedPath    string   `json:"referencedPath"` // encoded path of the referenced table (P7)
	ReferencedColumns []string `json:"referencedColumns"`
	OnDelete          *string  `json:"onDelete"`
	OnUpdate          *string  `json:"onUpdate"`
}

// ObjectMeta mirrors domain/tree.ts's objectMetaSchema.
type ObjectMeta struct {
	Path          string           `json:"path"`
	Kind          string           `json:"kind"`
	Name          string           `json:"name"`
	QualifiedName string           `json:"qualifiedName"`
	Columns       []ColumnMeta     `json:"columns"`
	PrimaryKey    []string         `json:"primaryKey"`
	ForeignKeys   []ForeignKeyMeta `json:"foreignKeys"`
	ReferencedBy  []ForeignKeyMeta `json:"referencedBy"` // D17
	Indexes       []IndexMeta      `json:"indexes"`
	RowEstimate   *int             `json:"rowEstimate"`
	Comment       *string          `json:"comment"`
}

// EncodePath mirrors tree.ts:33-35's encodePath: 'schema:public/table:order%2Fitems' — the
// connection id is not part of the string (D6).
func EncodePath(segments []PathSegment) string {
	parts := make([]string, len(segments))
	for i, s := range segments {
		parts[i] = s.Kind + ":" + EncodeURIComponent(s.Name)
	}
	return strings.Join(parts, "/")
}

// DecodePath ports tree.ts:37-49 exactly: "" -> no segments; split on '/'; each segment is
// <kind>:<encodeURIComponent(name)>; an unknown kind or a missing ':' is an error (the TS
// throws).
func DecodePath(connectionID, encoded string) (NodePath, error) {
	if encoded == "" {
		return NodePath{ConnectionID: connectionID, Segments: []PathSegment{}}, nil
	}
	raw := strings.Split(encoded, "/")
	segments := make([]PathSegment, len(raw))
	for i, seg := range raw {
		sep := strings.IndexByte(seg, ':')
		if sep < 0 {
			return NodePath{}, fmt.Errorf("model: malformed path segment: %s", seg)
		}
		kind := seg[:sep]
		if !ValidNodeKind(kind) {
			return NodePath{}, fmt.Errorf("model: unknown node kind in path segment: %s", kind)
		}
		segments[i] = PathSegment{Kind: kind, Name: DecodeURIComponent(seg[sep+1:])}
	}
	return NodePath{ConnectionID: connectionID, Segments: segments}, nil
}

// ValidateTreeNodes is the explicit check that replaces zod's safeParse for a cached []TreeNode
// row (P55 §1.6): every element's Kind must be a real NodeKind, and Name/Path non-empty.
func ValidateTreeNodes(nodes []TreeNode) bool {
	for _, n := range nodes {
		if !ValidNodeKind(n.Kind) || n.Name == "" || n.Path == "" {
			return false
		}
	}
	return true
}

// ValidateObjectMeta reports whether meta is well-formed, normalizing its four list fields from
// nil to [] first — json.Unmarshal leaves an absent JSON array as nil, not zod's own
// always-an-array guarantee, so a valid meta with an empty list must still come back usable.
func ValidateObjectMeta(meta *ObjectMeta) bool {
	if meta.Columns == nil {
		meta.Columns = []ColumnMeta{}
	}
	if meta.PrimaryKey == nil {
		meta.PrimaryKey = []string{}
	}
	if meta.ForeignKeys == nil {
		meta.ForeignKeys = []ForeignKeyMeta{}
	}
	if meta.ReferencedBy == nil {
		meta.ReferencedBy = []ForeignKeyMeta{}
	}
	if meta.Indexes == nil {
		meta.Indexes = []IndexMeta{}
	}
	return ValidNodeKind(meta.Kind) && meta.Path != "" && meta.Name != "" && meta.QualifiedName != ""
}
